### PR TITLE
add Cmd to pkg

### PR DESCRIPTION
**Pull request by Liujingfang1**
Add `Cmd` struct under pkg as the public library for apply, prune and delete.

Users can use this struct as the following
```
import sigs.k8s.io/cli-experimental/pkg
buf := new(bytes, Buffer)
c := InitializeCmd(buf, nil)
var resources []*unstructured.Unstructured
err := c.Apply(resources)
err = c.Prune(resources)
err = c.Delete(resources)
```

/assign @barney-s

### CODE REVIEW

File: pkg/wirepkg.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+var ProviderSet = wire.NewSet(
+	wire.Struct(new(apply.Apply), "DynamicClient", "Out"),
+	wire.Struct(new(prune.Prune), "DynamicClient", "Out"),
+	wire.Struct(new(delete.Delete), "DynamicClient", "Out"),
+	wirek8s.ProviderSet,
+	NewCmd,
+)
+
+// NewCmd returns a new Cmd object
+func NewCmd(a apply.Apply, p prune.Prune, d delete.Delete, c client.Client) (*Cmd, error) {
```
Comment (barney-s) [inline pkg/wirepkg.go]: Is this a sideeffect of wire. If we keep adding commands this list will grow. Any other way to structure this ?

Comment (Liujingfang1) [reply pkg/wirepkg.go]: I can remove this function.

**Comment by Liujingfang1** [comment]
@barney-s updated the change. PTAL

File: internal/pkg/prune/prune.go [hunk 2]
```diff
@@ -143,3 +143,28 @@ func (o *Prune) deleteObject(ctx context.Context, gvk schema.GroupVersionKind,
 	}
 	return obj, nil
 }
+
+// GetPruneResources finds the resource used for pruning from a slice of resources
+// by looking for a special annotation in the resource
+// inventory.InventoryAnnotation
+func GetPruneResources(resources []*unstructured.Unstructured) (*unstructured.Unstructured, error) {
```
Comment (seans3) [inline internal/pkg/prune/prune.go]: This function should have a unit test.

File: pkg/cmd_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	r2.SetName("cm2")
+	r2.SetNamespace("default")
+	r2.SetAnnotations(map[string]string{
+		inventory.InventoryAnnotation:     "{}",
+		inventory.InventoryHashAnnotation: "1234567",
+	})
+	return []*unstructured.Unstructured{r1, r2}
+}
+
+func TestCmd(t *testing.T) {
```
Comment (seans3) [inline pkg/cmd_test.go]: You probably want to add tests which exercise the error cases. What happens when you pass in nil to these three functions?

File: internal/pkg/prune/prune.go (hunk 2 above)
Comment (Liujingfang1) [reply internal/pkg/prune/prune.go]: I moved this function to `resourceconfig` package and added a unit test.

File: pkg/cmd_test.go (hunk 3 above)
Comment (Liujingfang1) [reply pkg/cmd_test.go]: Added a unit test for nil input.
Added a unit test for 2 consecutive applies.

File: internal/pkg/prune/prune.go [hunk 4]
```diff
@@ -143,3 +143,28 @@ func (o *Prune) deleteObject(ctx context.Context, gvk schema.GroupVersionKind,
...
+
+	for _, res := range resources {
+		annotations := res.GetAnnotations()
+		if _, ok := annotations[inventory.InventoryAnnotation]; ok {
+			count++
+			result = res
+		}
+	}
+
+	switch count {
```
[the remark this answers is not in the document]
Comment (Liujingfang1) [reply internal/pkg/prune/prune.go]: updated

File: pkg/cmd.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	a.Applier.Resources = resources
+	_, err := a.Applier.Do()
+	return err
+}
+
+// Prune prunes resources given the input as a slice of unstructured resources
+func (a *Cmd) Prune(resources []*unstructured.Unstructured) error {
+	pruneResource, err := prune.GetPruneResources(resources)
+	if err != nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (Liujingfang1) [reply pkg/cmd.go]: Good catch. Thank you :)

File: pkg/cmd.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	}
+	a.Pruner.Resources = pruneResource
+	_, err = a.Pruner.Do()
```
[the remark this answers is not in the document]
Comment (Liujingfang1) [reply pkg/cmd.go]: All those three functions will trigger a `Do` method. 
The parameters are passed when we initialize the `Cmd` object.
```
InitializeCmd(io.Writer, util.Args)
```
`io.Writer` stores the output. `util.Args` is for the config flags such as `--context`.

**Comment by Liujingfang1** [comment]
Updated tests and addressed comments. @seans3 @apelisse PTAL

File: internal/pkg/resourceconfig/resourceconfig.go [hunk 7]
```diff
@@ -177,3 +178,27 @@ func (p *RawConfigHTTPProvider) GetConfig(path string) ([]*unstructured.Unstruct
...
+
+	for _, res := range resources {
+		annotations := res.GetAnnotations()
+		if _, ok := annotations[inventory.InventoryAnnotation]; ok {
+			count++
+			result = res
+		}
+	}
+
+	if count == 0 {
```
Comment (seans3) [inline internal/pkg/resourceconfig/resourceconfig.go]: I don't think this if statement is needed, but it is more descriptive.

**Pull request merged**